plugin/plugin_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 the Drone Authors. All rights reserved.
+// Use of this source code is governed by the Blue Oak Model License
+// that can be found in the LICENSE file.
+
+package plugin
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+func fatalIf(err error) {
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func setupFilesAndFolders() string {
+	tempDir, err := os.MkdirTemp("", "fileglob")
+	fatalIf(err)
+
+	fatalIf(os.MkdirAll(filepath.Join(tempDir, "abc/def"), 0755))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/def/one.txt"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/def/one.yml"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/def/one.xml"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/def/two.txt"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/one.txt"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/one.yml"), []byte{}, 0644))
+	fatalIf(os.WriteFile(filepath.Join(tempDir, "abc/two.txt"), []byte{}, 0644))
+
+	return tempDir
+}
+
+func Test_validateArg_SunnyDay(t *testing.T) {
+	err := validateArgs(Args{
+		Filter: "**/*.txt",
```

Comment:
`**/.txt` and excludes `**/def/, /.yml`, `**/.xml` are added in the test, could we also add tests for single character match (`?`), mix of `*` and `?`, directory wildcards like `org/apache/**/CVS/*` or `**/test/**` as mentioned in https://ant.apache.org/manual/dirtasks.html

Reply:
I pushed for a review two more test cases ```Test_Exec_SingleCharacter``` and ```func Test_Exec_DirectoryWildcards``` but both failed.

Reply:
quickly added some debug statements to investigate the two failing test cases.

<details>
  <summary>here's the output from the tests:</summary>

=== RUN   Test_Exec_SingleCharacter
Current directory in Test_Exec_SingleCharacter: /private/var/folders/zv/dn6hzb0j06vc2_bsywqlmcd00000gn/T/fileglob1142222436
File a.xyz exists: true
File b.xyz exists: true
Matched files: []
    plugin_test.go:211:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:211
        	Error:      	Received unexpected error:
        	            	file does not exist
        	Test:       	Test_Exec_SingleCharacter
    plugin_test.go:212:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:212
        	Error:      	"[]" should have 2 item(s), but has 0
        	Test:       	Test_Exec_SingleCharacter
    plugin_test.go:218:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:218
        	Error:      	[]string(nil) does not contain "a.xyz"
        	Test:       	Test_Exec_SingleCharacter
    plugin_test.go:219:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:219
        	Error:      	[]string(nil) does not contain "b.xyz"
        	Test:       	Test_Exec_SingleCharacter
--- FAIL: Test_Exec_SingleCharacter (0.00s)
=== RUN   Test_Exec_DirectoryWildcards
Current directory in Test_Exec_DirectoryWildcards: /private/var/folders/zv/dn6hzb0j06vc2_bsywqlmcd00000gn/T/fileglob2150067361
File main.go exists: true
File go.mod exists: true
File go.sum exists: true
Matched files: [{community abc/test/harness/community true 160 2024-09-11T18:45:30+05:30}]
    plugin_test.go:242:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:242
        	Error:      	"[{community abc/test/harness/community true 160 2024-09-11T18:45:30+05:30}]" should have 3 item(s), but has 1
        	Test:       	Test_Exec_DirectoryWildcards
    plugin_test.go:248:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:248
        	Error:      	[]string{"abc/test/harness/community"} does not contain "abc/test/harness/community/main.go"
        	Test:       	Test_Exec_DirectoryWildcards
    plugin_test.go:249:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:249
        	Error:      	[]string{"abc/test/harness/community"} does not contain "abc/test/harness/community/go.mod"
        	Test:       	Test_Exec_DirectoryWildcards
    plugin_test.go:250:
        	Error Trace:	/Users/ompragash/Git/fileglobtest/plugin/plugin_test.go:250
        	Error:      	[]string{"abc/test/harness/community"} does not contain "abc/test/harness/community/go.sum"
        	Test:       	Test_Exec_DirectoryWildcards
--- FAIL: Test_Exec_DirectoryWildcards (0.00s)
</details>

for `Test_Exec_SingleCharacter`,  looks like the pattern ?.xyz is not matching a.xyz and b.xyz, even though the files exist. This indicates that the zglob library may not be handling the ? wildcard as expected
and for `Test_Exec_DirectoryWildcards`, pattern **/harness/**  is returning 1 directory (community), but the test is expecting to match the files main.go, go.mod, and go.sum. this pattern is only capturing the directory

Could you confirm if the zglob library supports Ant-style patterns, specifically for single characters with ? and recursive file matching (**/harness/**)

Reply:
Not all the Ant path style is supported by [zglob](https://pkg.go.dev/github.com/mattn/go-zglob) lib and needs to be replaced by [go-antpathmatcher](https://github.com/georgeJobs/go-antpathmatcher) that support that style.

It’s important to note that a pattern and a path must both be absolute or relative for the two to match, user cannot mix them.